src/testbase.rs

```diff
@@ -0,0 +1,300 @@
+
+        // It should hit exactly three boxes.
+        let mut expected_shapes = HashSet::new();
+        expected_shapes.insert(4);
+        expected_shapes.insert(5);
+        expected_shapes.insert(6);
+        traverse_and_verify(origin, direction, &all_shapes, &bh, &expected_shapes);
+    }
+}
+
+/// A triangle struct. Instance of a more complex `Bounded` primitive.
+pub struct Triangle {
+    pub a: Point3<f32>,
+    pub b: Point3<f32>,
+    pub c: Point3<f32>,
+    aabb: AABB,
+}
+
+impl Triangle {
+    fn new(a: Point3<f32>, b: Point3<f32>, c: Point3<f32>) -> Triangle {
+        Triangle {
+            a: a,
+            b: b,
+            c: c,
+            aabb: AABB::empty().grow(&a).grow(&b).grow(&c),
+        }
+    }
+}
+
+impl Bounded for Triangle {
+    fn aabb(&self) -> AABB {
+        self.aabb
+    }
+}
+
+/// Creates a unit size cube centered at `pos` and pushes the triangles to `shapes`.
+fn push_cube(pos: Point3<f32>, shapes: &mut Vec<Triangle>) {
+    let top_front_right = pos + Vector3::new(0.5, 0.5, -0.5);
+    let top_back_right = pos + Vector3::new(0.5, 0.5, 0.5);
+    let top_back_left = pos + Vector3::new(-0.5, 0.5, 0.5);
+    let top_front_left = pos + Vector3::new(-0.5, 0.5, -0.5);
+    let bottom_front_right = pos + Vector3::new(0.5, -0.5, -0.5);
+    let bottom_back_right = pos + Vector3::new(0.5, -0.5, 0.5);
+    let bottom_back_left = pos + Vector3::new(-0.5, -0.5, 0.5);
+    let bottom_front_left = pos + Vector3::new(-0.5, -0.5, -0.5);
+
+    shapes.push(Triangle::new(top_back_right, top_front_right, top_front_left));
+    shapes.push(Triangle::new(top_front_left, top_back_left, top_back_right));
+    shapes.push(Triangle::new(bottom_front_left, bottom_front_right, bottom_back_right));
+    shapes.push(Triangle::new(bottom_back_right, bottom_back_left, bottom_front_left));
+    shapes.push(Triangle::new(top_back_left, top_front_left, bottom_front_left));
+    shapes.push(Triangle::new(bottom_front_left, bottom_back_left, top_back_left));
+    shapes.push(Triangle::new(bottom_front_right, top_front_right, top_back_right));
+    shapes.push(Triangle::new(top_back_right, bottom_back_right, bottom_front_right));
+    shapes.push(Triangle::new(top_front_left, top_front_right, bottom_front_right));
+    shapes.push(Triangle::new(bottom_front_right, bottom_front_left, top_front_left));
+    shapes.push(Triangle::new(bottom_back_right, top_back_right, top_back_left));
+    shapes.push(Triangle::new(top_back_left, bottom_back_left, bottom_back_right));
+}
+
+/// Implementation of splitmix64.
+/// For reference see: http://xoroshiro.di.unimi.it/splitmix64.c
+fn splitmix64(x: &mut u64) -> u64 {
+    *x = x.wrapping_add(0x9E3779B97F4A7C15u64);
+    let mut z = *x;
+    z = (z ^ (z >> 30)).wrapping_mul(0xBF58476D1CE4E5B9u64);
+    z = (z ^ (z >> 27)).wrapping_mul(0x94D049BB133111EBu64);
+    z ^ (z >> 31)
+}
+
+/// Generates a new Point3, mutates the seed.
+fn next_point3(seed: &mut u64) -> Point3<f32> {
```

Comment:
The create_bih branch has a [different implementation](https://github.com/svenstaro/bvh/blob/create-bih/src/testbase.rs#L127). Which one is better?